Sources/SwiftOperatorPrecedence/OperatorPrecedence+Folding.swift

```diff
@@ -0,0 +1,364 @@
+        return try fold(
+          lhs, rest: &rest, bound: bound, errorHandler: errorHandler
+        )
+
+      case .none:
+        // If we ended up here, it's because we're either:
+        //   - missing precedence groups,
+        //   - have unordered precedence groups, or
+        //   - have the same precedence group with no associativity.
+        if let op1Precedence = op1Precedence,
+            let op2Precedence = op2Precedence {
+          try errorHandler(
+            .incomparableOperators(
+              leftOperator: op1, leftPrecedenceGroup: op1Precedence,
+              rightOperator: op2, rightPrecedenceGroup: op2Precedence
+            )
+          )
+        }
+
+        // Recover by folding arbitrarily at this operator, then continuing.
+        lhs = makeBinaryOperationExpr(lhs: lhs, op: op1, rhs: rhs)
+        return try fold(lhs, rest: &rest, bound: bound, errorHandler: errorHandler)
+      }
+    }
+
+    // Fold LHS and RHS together and declare completion.
+    return makeBinaryOperationExpr(lhs: lhs, op: op1, rhs: rhs)
+  }
+
+  /// "Fold" a sequence expression into a structured syntax tree.
+  ///
+  /// A sequence expression results from parsing an expression involving
+  /// infix operators, such as `x + y * z`. Swift's grammar does not
+  /// involve operator precedence, so a sequence expression is a flat list
+  /// of all of the terms `x`, `+`, `y`, `*`, and `z`. This operation folds
+  /// a single sequence expression into a structured syntax tree that
+  /// represents the same source code, but describes the order of operations
+  /// as if the expression has been parenthesized `x + (y * z)`.
+  public func foldSingle(
+    _ sequence: SequenceExprSyntax,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> ExprSyntax {
+    let lhs = sequence.elements.first!
+    var rest = sequence.elements.dropFirst()
+    return try fold(
+      lhs, rest: &rest,
+      bound: PrecedenceBound(groupName: nil, isStrict: false, syntax: nil),
+      errorHandler: errorHandler
+    )
+  }
+
+  /// Syntax rewriter that folds all of the sequence expressions it
+  /// encounters.
+  private class SequenceFolder : SyntaxRewriter {
+    /// The first operator precedecence that caused the error handler to
+    /// also throw.
+    var firstFatalError: OperatorPrecedenceError? = nil
+
+    let opPrecedence: OperatorPrecedence
+    let errorHandler: OperatorPrecedenceErrorHandler
+
+    init(
+      opPrecedence: OperatorPrecedence,
+      errorHandler: @escaping OperatorPrecedenceErrorHandler
+    ) {
+      self.opPrecedence = opPrecedence
+      self.errorHandler = errorHandler
+    }
+
+    override func visit(_ node: SequenceExprSyntax) -> ExprSyntax {
+      // If the error handler threw in response to an error, don't
+      // continue folding.
+      if firstFatalError != nil {
+        return ExprSyntax(node)
+      }
+
+      let newNode = super.visit(node).as(SequenceExprSyntax.self)!
```

Comment:
I don’t think this is necessary. The base `SyntaxRewriter` is a no-op.

Reply:
I needed it to visit the children, or else this transform wouldn't be recursive. Am I holding it wrong?

Reply:
Nevermind, I was stupid.